spec.bs

```diff
@@ -688,26 +690,62 @@ To <dfn>generate and score bids</dfn> given an [=auction config=] |auctionConfig
 To <dfn>score and rank a bid</dfn> given an [=auction config=] |auctionConfig|, a [=generated bid=]
 |generatedBid|, a [=leading bid info=] |leadingBidInfo|, a [=string=] |decisionLogicScript|, a
 {{unsigned long}}-or-null |biddingDataVersion|, and a [=boolean=] |isComponentAuction|:
-1. TODO: Assemble and fetch trusted scoring signals
-1. TODO: Let |adMetadata| be ... maybe include in [=generated bid=]
+1. Let |renderUrls| be a new [=list/is empty|empty=] [=list=].
+1. Let |adComponentRenderUrls| be a new [=list/is empty|empty=] [=list=].
+1. If |generatedBid|'s [=generated bid/ad descriptor=] is not null:
+  1. [=list/Append=] |generatedBid|'s [=generated bid/ad descriptor=]'s [=ad descriptor/url=] to
+    |renderUrls|.
+1. If |generatedBid|'s [=generated bid/ad component descriptors=] is not null:
+  1. [=list/For each=] |adComponentDescriptor| in |generatedBid|'s
+    [=generated bid/ad component descriptors=]:
+    1. [=list/Append=] |adComponentDescriptor|'s [=ad descriptor/url=] to |adComponentRenderUrls|.
+1. Let |fullSignalsUrl| be the result of [=building trusted scoring signals url=] with |auctionConfig|'s
+  [=auction config/trusted scoring signals url=], |renderUrls|, |adComponentRenderUrls|,
+  |auctionConfig|'s [=auction config/seller experiment group id=].
+
+  Note: Implementations can batch requests by collecting render URLs and ad component render URLs
+  from multiple invocations of [=score and rank a bid=] and passing them all to a single invocation
+  of [=building trusted scoring signals url=] -- the network response has to be parsed to pull out the pieces
+  relevant to each scoreAd() call.
+1. Let |trustedScoringSignals| be null.
+1. Let «|allTrustedScoringSignals|, |dataVersion|» be the result of [=fetching trusted signals=] with
+  |fullSignalsUrl| and false.
+1. If |allTrustedScoringSignals| is not failure, and if |allTrustedScoringSignals| is

```

Comment:
I think we only need to check if it's an [=ordered map=].
Same to the other comment for trusted bidding signals, [=fetching trusted signals=] does not return failure, but null when there's an error (of course we can change that if failure is better although I don't see a reason to do so).

Reply:
Done